enhancements/microshift/lvms-topolvm-downstream-replacement.md

```diff
@@ -0,0 +1,269 @@
+6. MicroShift removes the old TopoLVM deployment without removing the CRDs.
+   All existing storageClasses and deviceClasses are still present.
+7. MicroShift installs LVMS with a LVMCluster CustomResource.
+   * The LVMCluster is configured with `.spec.configurationPolicy=ExternalConfiguration`
+     which will trigger a deployment of the driver which can use the existing
+     deviceClasses and storageClasses.
+   * If LVMS does not detect a lvmd.yaml file (at runtime), it will error out,
+     as an external configuration is now mandatory for correct configuration.
+     Note that if the lvmd.yaml file is not present at install time, we should
+     keep the current behavior of not installing LVMS at all.
+   * When External Configuration is not used, then not specifying any deviceClasses
+     will result in rejection of the LVMCluster specification.
+   * Note that while it is eventually planned to decouple the CSI stack from MicroShifts
+     core resources, this enhancement only targets the migration of the existing solution,
+     not the adoption of LVMS as an external component out-of-tree.
+8. MicroShift continues to run, and LVMS starts the vg-manager DaemonSet with all CSI bindings included.
+9. All existing configurations continue to work
+   * Errors of TopoLVM in the node will now be written to vg-manager.
+   * Errors of the LVMCluster deployment will be written to the lvms-controller.
+
+### API Extensions
+
+MicroShift will now introduce the `LVMCluster`, `LVMVolumeGroup` and `LVMVolumeGroupNodeStatus` CustomResources.
+All existing documentation of LVMS on the CR usages applies and can be cross-referenced if users
+want a simplified setup of their storage devices.
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+N/A
+
+#### Standalone Clusters
+
+N/A
+
+#### Single-node Deployments or MicroShift
+
+This enhancement only applies to MicroShift.
+
+### Implementation Details/Notes/Constraints
+
+#### Enablement of configuration of no deviceClasses in LVMS
+
+Currently, LVMS does not allow for the configuration of no deviceClasses in the LVMCluster.
+This is a constraint that we will have to work around by allowing the user to specify
+an empty deviceClass in the LVMCluster and then coordinating that with the vg-manager.
+We have a prototype implementation that can be pushed into GA in time for the 4.17 release based on
+[OCPEDGE-830](https://issues.redhat.com//browse/OCPEDGE-830) that proves this is possible.
+
+#### Switching from TopoLVM to LVMS
+
+The switch from TopoLVM to LVMS will be done in a single release. We will support
+upgrading from 4.16 to 4.16, and from 4.14 to 4.16 (EUS).
+
+The switch will be done by removing the TopoLVM deployment and installing the LVMS
+deployment. The LVMS deployment will be configured to use the existing deviceClasses
+and storageClasses. There will be leftover images from the TopoLVM deployment that
+can be removed manually from the registry after the upgrade.
```

Comment:
Are we planning to deploy a kustomization manifest? Can it take care of image removal as well to free up space?

Reply:
I do not believe we have anything to clean up unused images. I will ask the team to comment with more ideas here.

Reply:
cri-o and the kubelet have [garbage collection](https://kubernetes.io/docs/concepts/architecture/garbage-collection/#containers-images). I don't think we need to do anything explicitly.

Reply:
Reworded this paragraph to make clear that the leftover images will be tackled through GC.